R/mlaps.R

```diff
@@ -0,0 +1,395 @@
+#' Since not all encounters are admitted through Emergency Department, depending on research question, it can be relevant to consider lab tests collected in early inpatient admission.
+#' Typically, `hours_after_admission` can be set to 24 to consider any lab tests collected at Emergency Department and 24 hours after inpatient admission.
+#'
+#' @param output_laps_components (`logical`)\cr
+#' Does not aggregate the score and instead outputs for each LAPS component (test), its contribution
+#' to the overall score.
+#'
+#' @return (`data.frame`)\cr
+#' If `output_laps_components == TRUE`:
+#'     `genc_id` (`numeric`),\cr
+#'     `test_type_mapped_omop` (`character`),\cr
+#'     `mlaps` (`numeric`) max score for this test.
+#'
+#' If `output_laps_components == FALSE`:
+#'     `genc_id` (`numeric`),\cr
+#'     `mlaps` (`numeric`) sum of max scores for each relevant test for this encounter.
+#'
+#' @import DBI RPostgreSQL
+#' @importFrom purrr map2_df
+#' @export
+#'
+#' @references
+#' When the function is used, please cite the following:
+#' https://doi.org/10.1097/MLR.0b013e3181589bb6
+#' https://doi.org/10.1007/s11606-023-08245-w
+#' https://doi.org/10.1101/2023.01.06.23284273
+#'
+loop_mlaps <- function(db, cohort = NULL, hours_after_admission = 0, output_laps_components = FALSE) {
+
+  hospital_field <- return_hospital_field(db)
+
+  admdad <- DBI::dbGetQuery(
+    db,
+    paste(
+      "SELECT
+        genc_id,
+        admission_date_time,
+        EXTRACT(YEAR FROM discharge_date_time::DATE) AS year,",
+        hospital_field, "AS hospital_id",
+      "FROM admdad",
+      if (!is.null(cohort)) {
+        paste("WHERE genc_id IN (", paste(cohort, collapse = ", "), ")")
+      }
+    )
+  ) %>%
+    as.data.table()
+
+  hospital_years <- admdad %>%
+    select(hospital_id, year) %>%
+    unique()
+
+  res <- purrr::map2_df(
+    hospital_years$hospital_id,
+    hospital_years$year,
+    function(hospital_id, year) {
+      lab <- DBI::dbGetQuery(
+        db,
+        paste(
+          "SELECT
+            l.genc_id,
+            l.collection_date_time,
+            l.test_type_mapped_omop,
+            l.result_value,
+            l.result_unit
+          FROM lab l
+          INNER JOIN admdad a
+            ON l.genc_id = a.genc_id
+          WHERE l.test_type_mapped_omop IN (", paste(LAPS_OMOP_CONCEPTS, collapse = ", "), ")",
+          paste0("AND a.", hospital_field, " = '", hospital_id, "'"),
+          "AND EXTRACT(YEAR FROM a.discharge_date_time::DATE) = ", year,
+          if (!is.null(cohort)) {
+            paste("AND l.genc_id IN (", paste(cohort, collapse = ", "), ")")
+          }
+        )
+      ) %>%
+        as.data.table()
+
+      res <- mlaps(
+        admdad %>% dplyr::filter(hospital_id == hospital_id, year == year),
+        lab,
+        hours_after_admission = hours_after_admission,
+        componentwise = output_laps_components
+      )
+    }
+  )
+
+  return(res)
+}
+
+
+#' @title
+#' mLAPS
+#'
+#' @details
```

Comment:
Not sure what the best spot for this is, but can you add somewhere in the documentation how the function handles scenarios where multiple results are found for the same lab tests (looks like the max result is used except for creatinine?)?